src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -49,41 +57,142 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
   function updateFunctionalComponent(current, workInProgress) {
     var fn = workInProgress.type;
     var props = workInProgress.pendingProps;
-    console.log('update fn:', fn.name);
     var nextChildren = fn(props);
     reconcileChildren(current, workInProgress, nextChildren);
     workInProgress.pendingWorkPriority = NoWork;
   }
 
-  function updateHostComponent(current, workInProgress) {
-    console.log(
-      'host component', workInProgress.type, 
-      typeof workInProgress.pendingProps.children === 'string' ? workInProgress.pendingProps.children : ''
-    );
+  function scheduleUpdate(fiber: Fiber, pendingState: PendingState, priorityLevel : PriorityLevel): void {
+    const { scheduleLowPriWork } = getScheduler();
+    fiber.pendingState = pendingState;
+    while (true) {
+      if (fiber.pendingWorkPriority === NoWork ||
+          fiber.pendingWorkPriority >= priorityLevel) {
+        fiber.pendingWorkPriority = priorityLevel;
+      }
+      // Duck type root
+      if (fiber.stateNode && fiber.stateNode.containerInfo) {
+        const root : FiberRoot = (fiber.stateNode : any);
+        scheduleLowPriWork(root, priorityLevel);
+        return;
+      }
+      if (!fiber.return) {
+        throw new Error('No root!');
+      }
+      fiber = fiber.return;
+    }
+  }
+
+  // Class component state updater
+  const updater = {
+    enqueueSetState(instance, partialState) {
+      const fiber = instance._fiber;
+      let pendingState = fiber.pendingState;
+
+      // Append to pending state queue
+      const nextPendingStateNode = createPendingState(partialState);
+      if (pendingState === null) {
+        pendingState = nextPendingStateNode;
+      } else {
+        if (pendingState.tail === null) {
+          pendingState.next = nextPendingStateNode;
+        } else {
+          pendingState.tail.next = nextPendingStateNode;
+        }
+        pendingState.tail = nextPendingStateNode;
+      }
+
+      // Must schedule an update on both alternates, because we don't know tree
+      // is current.
+      scheduleUpdate(fiber, pendingState, LowPriority);
+      if (fiber.alternate) {
+        scheduleUpdate(fiber.alternate, pendingState, LowPriority);
```

Comment:
The parent path is guaranteed to be the same conceptual component path. However, you can't rely on any particular node being the current one along the path.

So this is safe:

```js
while (node) {
  addState(node);
  addState(node.alternate);
  node = node.return;
}
```

But this is not safe, because it is not guaranteed to cover every node:

```js
node = start;
while (node) {
  addState(node);
  node = node.return;
}
node = start.alternate;
while (node) {
  addState(node);
  node = node.return;
}
```

An example when this happens:

```js
// Assume the cycles are resolved
a1 = { child: b1, return: null, alternate: a2 };
a2 = { child: b2, return: null, alternate: a1 };
b1 = { child: c1, return: a1, alternate: b2 };
b2 = { child: c1, return: a2, alternate: b1 };
c1 = { child: null, return: b1, alternate: c2 };
c2 = { child: null, return: b1, alternate: c1 }; // Note that the return is the same b1.
```

Since two versions of a fiber can point to the same child, it is not possible for a parent pointer alone to point to every parent since there can be multiple.

Conceptually this pointer is used for two purposes. A pointer to the parent "Instance" which I've merged into the Fiber. See the type definition.

However, it is also used as a temporary pointer for knowing which fiber to step back through after processing.

If I wasn't so unnecessarily clever about allocations the structure would like this:

```js
type Instance = { parent: ?Instance };
type Fiber = { inst: Instance, return: ?Fiber };
```

and you would just walk the parent pointer. However, I'm being clever to see how far we can get but it adds some complexity.